gh-99108: Import SHA2-224 and SHA2-256 from HACL*

msprotz opened the pull request:
Issue #99108 is about replacing hashlib primitives (for the non-OpenSSL case) with verified implementations from HACL*. This is the first PR in the series, and focuses specifically on SHA2-256 and SHA2-224.

This PR imports Hacl_Streaming_SHA2 into the Python tree. This is the HACL* implementation of SHA2, which combines a core implementation of SHA2 along with a layer of buffer management that allows updating the digest with any number of bytes. This supersedes the previous implementation in the tree.

@franziskuskiefer was kind enough to benchmark the changes: in addition to being verified (thus providing significant safety and security improvements), this implementation also provides a sizeable performance boost!

```
---------------------------------------------------------------
Benchmark                     Time             CPU   Iterations
---------------------------------------------------------------
Sha2_256_Streaming            3163 ns      3160 ns       219353     // this PR
LibTomCrypt_Sha2_256          5057 ns      5056 ns       136234     // library used by Python currently
``` 

The changes in this PR are as follows:
- import the subset of HACL* that covers SHA2-256/224 into `Modules/_hacl`
- rewire sha256module.c to use the HACL* implementation

To review the changes, one should focus on `sha256module.c`, and possibly `Hacl_Streaming_SHA2.h` which is the API entry point. The rest of the files are part of HACL* and are cherry-picked straight from the HACL* repository, so as to minimize the amount of hand-edits and facilitate future refreshes of the code, should HACL* land some performance improvements.

The original code comes from https://github.com/project-everest/hacl-star/tree/master/dist/gcc-compatible. The various .h files in `include/`, `krmllib/`, etc. will be shared across (hopefully) future primitives imported from HACL*. I can trim those files to minimize the amount of includes, but then this will make refreshing the code in the future harder (since the trimming will have to be redone for each refresh).

The code was originally authored by @karthikbhargavan and I; @polubelova and @r1km provided a considerable amount of help over the past week, as we were overhauling our implementation to offer a significant performance boost (we want this first PR to be a good one!). Thanks to everyone involved in this team effort.

There are a couple remarks left in the source code, but for now, I think it's better to open up a PR and start discussing. As promised, tagging @alex for this PR. Thanks so much!

<!-- gh-issue-number: gh-99108 -->
* Issue: gh-99108
<!-- /gh-issue-number -->

msprotz committed:
Replace SHA2 implementation with verified code from HACL*

msprotz committed:
Fixup some casts

msprotz committed:
And proper error handling for maximum hashing lengths exceeded

msprotz committed:
Remove extra file

msprotz committed:
See if using srcdir helps

msprotz committed:
Delete 11 un-needed files

msprotz committed:
Wrong direction for error checking

msprotz committed:
Remove another four files

msprotz committed:
better code quality

msprotz committed:
Fixup memory management mistake

msprotz committed:
Fix Linux build

gpshead reviewed (commented):
It'd be useful for this kind of thing to include a script (somewhere under Tools?.. where to put it can be figured out later) that can generate and update the hacl-star vendored code implementation. As I assume we'll not want this to be a one time code drop but periodically update it, at least for every major Python release, as hacl-star evolves.

tiran commented on Modules/Setup.stdlib.in:
```diff
@@ -78,7 +78,7 @@
 @MODULE__MD5_TRUE@_md5 md5module.c
 @MODULE__SHA1_TRUE@_sha1 sha1module.c
-@MODULE__SHA256_TRUE@_sha256 sha256module.c
+@MODULE__SHA256_TRUE@_sha256 _hacl/sha256module.c _hacl/Hacl_Streaming_SHA2.c -I$(srcdir)/Modules/_hacl/include -I$(srcdir)/Modules/_hacl/internal -I$(srcdir)/Modules/_hacl/krmllib/dist/minimal -DKRML_VERIFIED_UINT128 -D_BSD_SOURCE -D_DEFAULT_SOURCE
```
Includes and defines belong into `PY_STDLIB_MOD` block for sha256module in `configure.ac`.

msprotz replied:
Thanks for pointing me to the right location... I've fixed this, regenerated the configure script using autoconf-2.69 (to minimize the diff), and also moved sha256module.c (and the clinic-generated file) back into their original locations.

tiran reviewed (commented):
Please add all new headers to `MODULE__SHA256_DEPS` in `Makefile.pre.in`.

I would prefer if you keep the `sha256module.c` file in its current place and only replace libtom code with HACL calls. It's easier to review and keeps the history for untouched lines.

msprotz committed:
Address review comments from @tiran; regenerate configure

msprotz committed:
Merge remote-tracking branch 'origin/sha2_hacl' into sha2_hacl

msprotz committed:
Move Linux & BSD CFLAGS into configure

msprotz committed:
Add NEWS entry.

msprotz committed:
SRCDIRS

msprotz committed:
Properly regenerate configure

msprotz committed:
Fix paths in Windows build

msprotz committed:
Add include directory

msprotz committed:
Fixup include paths

msprotz committed:
Fix patchcheck

gpshead commented on Modules/sha256module.c:
```diff
@@ -67,321 +71,19 @@ _sha256_get_state(PyObject *module)
-    digest[29] = (unsigned char) ((sha_info->digest[7] >> 16) & 0xff);
-    digest[30] = (unsigned char) ((sha_info->digest[7] >>  8) & 0xff);
-    digest[31] = (unsigned char) ((sha_info->digest[7]      ) & 0xff);
+    // TODO: implement this in HACL
```
If there is an upstream issue on the HACL* repo, drop a link to that here.

msprotz replied:
implemented upstream for the purposes of this PR, and fixed in a632c84b6c09e87b51378821f171f1ddce8df6aa

gpshead commented on Modules/sha256module.c:
```diff
@@ -503,7 +200,8 @@ SHA256Type_update(SHAobject *self, PyObject *obj)
     GET_BUFFER_VIEW_OR_ERROUT(obj, &buf);
 
-    sha_update(self, buf.buf, buf.len);
+    if (Hacl_Streaming_SHA2_update_256(self->state, buf.buf, buf.len) != 0)
```
buf.len is a 64-bit Py_ssize_t, the HACL* method takes a uint32_t.  (convenient github actions compiler warning)

A loop around update seems necessary. And we need to ensure we've got a bigmem test covering a >4GiB input to update.

gpshead replied:
Related: We don't want to raise a RuntimeError as our API allows an arbitrary Python length input.

gpshead replied:
this also applies to the update calls in the constructors below.

msprotz replied:
what should be the behavior if the maximum hash length is exceeded? sha2-256 is only defined for inputs of length up to 2^63-1, and undefined otherwise

gpshead replied:
CPython isn't currently supported on any platform where `Py_ssize_t` could be larger than that $(2^{63} - 1)$ so I just wouldn't worry about it.

Have we ever had any code enforce that total sum of all update() method calls length in the past. Does anyone? It's an unreasonably long linear computation.

msprotz replied:
Regarding handling of data > 4GB: fixed with a small helper, and a suitable test in f4b267ebe24d5b8c33001b11eaab68a8f10e575c that feeds 4GB of zeroes, followed by "hello world" into the hash function.

Regarding checking the return code: I mis-spoke, the max length is 2^61-1. But in any case, even with best-in-class accelerators, it would take over a billion years to have that much data, so I think it's ok to ignore the return value -- this is what I did in the revised version.

msprotz committed:
Automate import of HACL* into the tree with a script

msprotz committed:
Use newly-implemented copy operation

msprotz committed:
Merge remote-tracking branch 'upstream/main' into sha2_hacl

msprotz committed:
Add a suitable test for data > 4GB

gpshead committed:
Make more robust & follow shellcheck.

* Require bash >= 4.
* Use arrays and proper quoting to avoid shell injection.
* Add more helpful help text and error messages.
* Update to the latest HACL* revision.

gpshead committed:
Update to the latest HACL* revision.

Which seems to just be a copyright message update.

gpshead committed:
Add a README to the _hacl directory.

gpshead committed:
reword and ReSTify the news entry.

gpshead committed:
Dynamically rename the Hacl_ C symbols.

We do this with other third party code as well, it avoids linking and
dynamic linking ODR violations when extension module or embedding
applications also use their own version of the library code in question.

+ Minor annotation on the test, use `python -m test -u cpu`
+ Add a link to the readme.

gpshead committed:
missing file add & readme update.

gpshead committed:
_

gpshead committed:
remove wrongly placed defines.

gpshead committed:
Merge branch 'main' into sha2_hacl

erlend-aasland commented on Modules/sha256module.c:
```diff
@@ -524,12 +228,13 @@ SHA256_get_block_size(PyObject *self, void *closure)
         return PyUnicode_FromStringAndSize("sha224", 6);
+    } else {
+        return PyUnicode_FromStringAndSize("sha256", 6);
+    }
```
```suggestion
    }
    return PyUnicode_FromStringAndSize("sha256", 6);
```

erlend-aasland reviewed (commented):
One small style nit. Also, there's a cast warning on line 127 in Modules/sha256module.c.

Autoconf changes looks good to me.

msprotz committed:
Stylistic fix in Modules/sha256module.c

Co-authored-by: Erlend E. Aasland <erlend.aasland@protonmail.com>

msprotz committed:
Fix a warning on Windows builders

gpshead reviewed (approved):
This PR is in good shape!  But on its own - lets call this step "2a" - it'll be a little odd to ship in 3.12 if we don't also have some natural related followups to include as well:

 * [x] 2b) SHA384 & SHA512 support (sha512module.c and related hacl code).
 * [ ] 3) SHA3 support (more generated hacl code).

**I'll go ahead and merge this**, and be responsible for reverting it if we decide we don't wind up just the shorter half of SHA2 using it if the follow-ons aren't ready yet before 3.12-beta1 in May.

Could you go ahead and prepare follow-on PRs?
